File: src/main/java/com/barcelos/recrutamento/core/service/BeneficioService.java
package com.barcelos.recrutamento.core.service;

import com.barcelos.recrutamento.core.exception.ResourceNotFoundException;
import com.barcelos.recrutamento.core.model.BeneficioOrg;
import com.barcelos.recrutamento.core.port.BeneficioOrgRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
public class BeneficioService {

    private final BeneficioOrgRepository beneficioOrgRepository;

    public BeneficioService(BeneficioOrgRepository beneficioOrgRepository) {
        this.beneficioOrgRepository = beneficioOrgRepository;
    }

    @Transactional
    public BeneficioOrg criar(UUID organizacaoId, String nome, String descricao) {
        var beneficio = BeneficioOrg.novo(organizacaoId, nome, descricao);
        return beneficioOrgRepository.save(beneficio);
    }

    @Transactional(readOnly = true)
    public BeneficioOrg buscar(UUID beneficioId) {
        return beneficioOrgRepository.findById(beneficioId)
                .orElseThrow(() -> new ResourceNotFoundException("Benefício", beneficioId));
    }

    @Transactional(readOnly = true)
    public List<BeneficioOrg> listarPorOrganizacao(UUID organizacaoId) {
        return beneficioOrgRepository.listByOrganizacao(organizacaoId);
    }

    @Transactional
    public BeneficioOrg atualizar(BeneficioOrg beneficio, String nome, String descricao) {
        var beneficioAtualizado = BeneficioOrg.atualizar(beneficio.getId(), beneficio.getOrganizacaoId(), nome, descricao);
        return beneficioOrgRepository.save(beneficioAtualizado);
    }

    @Transactional
    public void deletar(UUID beneficioId) {
        var beneficio = buscar(beneficioId);
        beneficioOrgRepository.deleteById(beneficioId);
    }
}

File: frontend/recrutamento/src/services/beneficio.service.ts
import api from "@/lib/api";

export interface Beneficio {
  id: string;
  nome: string;
  descricao: string;
  organizacaoId: string;
}

export interface CriarBeneficioRequest {
  nome: string;
  descricao: string;
}

export interface AtualizarBeneficioRequest {
  nome: string;
  descricao: string;
}

export const beneficioService = {
  async criar(
    organizacaoId: string,
    data: CriarBeneficioRequest
  ): Promise<Beneficio> {
    const response = await api.post(
      `/organizacoes/${organizacaoId}/beneficios`,
      data
    );
    return response.data;
  },

  async listarPorOrganizacao(organizacaoId: string): Promise<Beneficio[]> {
    const response = await api.get(`/organizacoes/${organizacaoId}/beneficios`);
    return response.data;
  },

  async buscar(organizacaoId: string, beneficioId: string): Promise<Beneficio> {
    const response = await api.get(
      `/organizacoes/${organizacaoId}/beneficios/${beneficioId}`
    );
    return response.data;
  },

  async atualizar(
    organizacaoId: string,
    beneficioId: string,
    data: AtualizarBeneficioRequest
  ): Promise<Beneficio> {
    const response = await api.put(
      `/organizacoes/${organizacaoId}/beneficios/${beneficioId}`,
      data
    );
    return response.data;
  },

  async deletar(organizacaoId: string, beneficioId: string): Promise<void> {
    await api.delete(
      `/organizacoes/${organizacaoId}/beneficios/${beneficioId}`
    );
  },

  async listarBeneficiosDaVaga(vagaId: string): Promise<string[]> {
    const response = await api.get(`/vagas/${vagaId}/beneficios`);
    return response.data;
  },
};
